### PR TITLE
Fix mid-epoch warning call while resuming

### LINKED ISSUE #11029: Resuming training throws the mid-epoch warning everytime
## Proposed refactor

Getting this:
```
UserWarning: You're resuming from a checkpoint that ended mid-epoch. Training will start from the beginning of the next epoch. This can cause unreliable results if further training is done, consider using an end of epoch checkpoint.
```
even when checkpoints saved at epoch end are being used to resume the training.

The reason is we set total train batches to `inf` here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/5576fbc5f9a7d0bc71ad26b8b54775110e675808/pytorch_lightning/trainer/trainer.py#L647

and reload dataloaders within fit_loop here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/5576fbc5f9a7d0bc71ad26b8b54775110e675808/pytorch_lightning/loops/fit_loop.py#L190-L193

so, num_training_batches is always `inf` as this point always.
https://github.com/PyTorchLightning/pytorch-lightning/blob/5576fbc5f9a7d0bc71ad26b8b54775110e675808/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L246-L253

### Pitch

Either remove the warning, since doesn't seem to resolve with the current logic
, or start adding a flag in all the checkpoints being saved indicating whether it was saved mid-epoch or not.
or any better solutions??

Else it will lead to false-positive warnings for users.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @ananthsub @ninginthecloud

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #11029

since we don't have the info regarding num training steps during checkpoint reload, we need to move it further once dataloaders are loaded.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: pytorch_lightning/loops/fit_loop.py [hunk 1]
```diff
@@ -195,6 +197,22 @@ def on_run_start(self) -> None:  # type: ignore[override]
         """Calls the ``on_train_start`` hook."""
         # reset train dataloader and val dataloader
         self.trainer.reset_train_val_dataloaders(self.trainer.lightning_module)
+
+        # global_step is incremented during checkpointing (#11555)
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:201]: Shouldn't this comment be in L207 instead?

File: pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -195,6 +197,22 @@ def on_run_start(self) -> None:  # type: ignore[override]
...
+        if self.restarting and self.trainer.num_training_batches not in (0, float("inf")):
+            self.trainer.accumulate_grad_batches = self.trainer.accumulation_scheduler.get_accumulate_grad_batches(
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:203]: Don't we need to set this attribute unconditionally?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -246,21 +246,6 @@ def restore_loops(self) -> None:
                 f" but you have set Trainer(max_epochs={self.trainer.max_epochs})."
             )
 
-        # Division deals with global step stepping once per accumulated batch
-        # Inequality deals with different global step for odd vs even num_training_batches
-        self.trainer.accumulate_grad_batches = self.trainer.accumulation_scheduler.get_accumulate_grad_batches(
-            self.trainer.current_epoch
-        )
-        n_accum = 1 if self.trainer.accumulate_grad_batches is None else self.trainer.accumulate_grad_batches
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:254]: Is this not necessary anymore? Does `accumulate_grad_batches` return always 1?

File: pytorch_lightning/loops/fit_loop.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:201]: yes!

File: pytorch_lightning/loops/fit_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:203]: sorry, didn't get you.

This is updated at every epoch. Even if I remove this line it will still work, but just kept it to be extra sure.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:254]: GradientScheduler returns 1 by default

File: pytorch_lightning/loops/fit_loop.py [hunk 4]
```diff
@@ -195,6 +197,22 @@ def on_run_start(self) -> None:  # type: ignore[override]
...
+                self.trainer.current_epoch
+            )
+            expected_steps = math.ceil(self.trainer.num_training_batches / self.trainer.accumulate_grad_batches)
+
+            if (self.trainer.global_step - 1) % expected_steps != 0:
```
Comment (rohitgr7) [inline pytorch_lightning/loops/fit_loop.py:208]:
```suggestion
        if self.restarting and self.trainer.num_training_batches not in (0, float("inf")):
            self.trainer.accumulate_grad_batches = self.trainer.accumulation_scheduler.get_accumulate_grad_batches(
                self.trainer.current_epoch
            )
            expected_steps = math.ceil(self.trainer.num_training_batches / self.trainer.accumulate_grad_batches)

            # global_step is incremented during checkpointing (#11555)
            if (self.trainer.global_step - 1) % expected_steps != 0:
```

File: pytorch_lightning/loops/fit_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:203]: I mean if we should do

```python
self.trainer.acccumulate_grad_batches = ...
if self.restarting and ...:
    ...
```

Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:203]: we need the value under certain conditions. if the condition doesn't satisfy, computing it outside won't do any good.

File: pytorch_lightning/loops/fit_loop.py [hunk 5]
```diff
@@ -195,6 +197,22 @@ def on_run_start(self) -> None:  # type: ignore[override]
         """Calls the ``on_train_start`` hook."""
         # reset train dataloader and val dataloader
         self.trainer.reset_train_val_dataloaders(self.trainer.lightning_module)
+
+        # global_step is incremented during checkpointing (#11555)
+        if self.restarting and self.trainer.num_training_batches not in (0, float("inf")):
```
Comment (tchaton) [inline pytorch_lightning/loops/fit_loop.py:202]: We should not display this warning is fault tolerant was used there.

Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py]: done!

File: pytorch_lightning/loops/fit_loop.py [hunk 6]
```diff
@@ -195,6 +198,23 @@ def on_run_start(self) -> None:  # type: ignore[override]
...
+            )
+            expected_steps = math.ceil(self.trainer.num_training_batches / self.trainer.accumulate_grad_batches)
+
+            # global_step is incremented during checkpointing (#11555)
+            if (self.trainer.global_step - 1) % expected_steps != 0:
+                rank_zero_warn(
+                    "You're resuming from a checkpoint that ended mid-epoch."
+                    " Training will start from the beginning of the next epoch."
+                    " This can cause unreliable results if further training is done,"
+                    " consider using an end of epoch checkpoint."
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:215]: This should instead suggest enabling fault-tolerance for full reproducibility.

#8578 will make it so we do not start from the beginning of the next epoch whether fault-tolerance is enabled or not

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/fit_loop.py (hunk 6 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:215]:
```suggestion
                    "You're resuming from a checkpoint that ended mid-epoch."
                    " Training will start from the beginning of the next epoch."
                    " This can cause unreliable results if further training is done,"
                    " consider using an end of epoch checkpoint or use fault-tolerant training"
                    " to enable full-reproducibility."
```

File: pytorch_lightning/loops/fit_loop.py [hunk 7]
```diff
@@ -195,6 +198,24 @@ def on_run_start(self) -> None:  # type: ignore[override]
...
+            expected_steps = math.ceil(self.trainer.num_training_batches / self.trainer.accumulate_grad_batches)
+
+            # global_step is incremented during checkpointing (#11555)
+            if (self.trainer.global_step - 1) % expected_steps != 0:
+                rank_zero_warn(
+                    "You're resuming from a checkpoint that ended mid-epoch."
+                    " Training will start from the beginning of the next epoch."
+                    " This can cause unreliable results if further training is done,"
+                    " consider using an end of epoch checkpoint or use fault-tolerant training"
+                    " to enable full-reproducibility."
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:216]:
```suggestion
                    " to restart as if training did not stop."
```

**Review by kaushikb11 — APPROVED** [review]

**Review by Borda — APPROVED** [review]
